File: scripts/curl.js
const fs = require("fs/promises");
const crypto = require("crypto");

function createReq(reqType, userId, once = true, cookieStr = undefined) {
  if (reqType === "chzzk_live") {
    const req = { uid: userId, once: once };
    if (cookieStr) req["cookies"] = cookieStr;
    return { reqType: reqType, chzzkLive: req };
  } else if (reqType === "afreeca_live") {
    return { reqType: reqType, afreecaLive: { userId: userId, once: once } };
  } else {
    throw new Error(`Unknown request type: ${reqType}`);
  }
}

function decrypt(encryptedText, key) {
  if (key.length !== 32) throw new Error("Key must be 32 bytes");
  const inputBuffer = Buffer.from(encryptedText, "base64");
  const decipher = crypto.createDecipheriv("aes-256-cfb", Buffer.from(key), inputBuffer.subarray(0, 16));
  const decrypted = Buffer.concat([decipher.update(inputBuffer.subarray(16)), decipher.final()]);
  return decrypted.toString("utf8");
}

(async () => {
  const [url, reqType, userId, hasCookie] = process.argv.slice(2);

  // get cookie string
  let cookieStr = undefined;
  if (hasCookie === "true") {
    const authed = JSON.parse(await fs.readFile("../dev/curl.json", 'utf-8'))["authed"];
    cookieStr = decrypt((await (await fetch(authed["url"])).json())["encrypted"], authed["enckey"]);
  }

  // send request
  const res = await fetch(url, {
    method: "POST", headers: {"Content-Type": "application/json"},
    body: JSON.stringify(createReq(reqType, userId, true, cookieStr)),
  });
  console.log(await res.text())
})();

File: scripts/curl.py
import json
import sys
import http.client


def create_req(req_type: str, userid: str, once: bool = True):
    if req_type == "chzzk_live":
        return {
            "reqType": req_type,
            "chzzkLive": {
                "uid": userid,
                "once": once,
            }
        }
    elif req_type == "afreeca_live":
        return {
            "reqType": req_type,
            "afreecaLive": {
                "userId": userid,
                "once": once,
            }
        }
    else:
        raise ValueError(f"Unknown request type: {req_type}")


def request(host: str, data: dict, path: str = "/stdl"):
    conn = http.client.HTTPConnection(host)
    conn.request(
        "POST", path, json.dumps(data),
        {"Content-Type": "application/json"}
    )
    return conn.getresponse()


if __name__ == "__main__":
    req = create_req(sys.argv[2], sys.argv[3], True)
    res = request(sys.argv[1], req)
    # res = request(sys.argv[1], req, "/stdl-proton")
    print(f"{res.status} {res.reason}")
